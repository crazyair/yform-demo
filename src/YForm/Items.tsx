import React, { useContext } from 'react';
import classNames from 'classnames';
import warning from 'warning';
import { find, set, merge, forEach, isObject, isArray, pick } from 'lodash';
import { FormInstance, FormItemProps } from 'antd/lib/form';

import { YFormContext, YFormItemsContext } from './Context';
import { replaceMessage, getLabelLayout } from './utils';
import { YFormPluginsType, YFormProps } from './Form';
import { YFormItemsTypeArray } from './ItemsType';
import ItemChildren from './ItemChildren';

export type YFormDataSource = YFormItemsTypeArray<YFormItemProps>;
export type YFormRenderChildren = (form: FormInstance) => YFormItemProps['children'];

export interface YFormItemProps extends Omit<FormItemProps, 'children'> {
	isShow?: boolean;
	required?: boolean;
	plugins?: YFormPluginsType | boolean;
	className?: string;
	addonAfter?: React.ReactElement;
	style?: React.CSSProperties;
	offset?: number;
	children?: (YFormDataSource | YFormDataSource[] | boolean)[] | React.ReactElement | YFormRenderChildren | boolean;
	dataSource?: YFormItemProps['children'];
}

export interface FormatFieldsValue<T = any> {
	name: FormItemProps['name'];
	format: (value: T) => any;
}

// 内部使用，类型不重要
interface InternalYFormItemProps extends YFormItemProps {
	dataSource?: any;
	items?: any;
	component?: any;
}

export interface YFormItemsProps extends YFormProps {
	offset?: number;
	noStyle?: boolean;
}

const Items = (props: YFormItemsProps) => {
	let defaultPlugin = true;
	const formProps = useContext(YFormContext);
	const itemsProps = useContext(YFormItemsContext);
	const { itemsType } = formProps;
	const { children = [], className, style, noStyle } = props;
	if ('isShow' in props && !props.isShow) return null;
	const mergeProps = merge({}, formProps, itemsProps, props);
	const { required: mergeRequired, disabled: mergeDisabled } = mergeProps;
	const list: React.ReactNode[] = [];
	const each = (lists: YFormItemsTypeArray<InternalYFormItemProps>[], pIndex?: number) => {
		forEach(lists, (item, index: number) => {
			const _index = pIndex ? `${pIndex}_${index}` : index;
			if (isArray(item)) {
				return each(item, index);
			}
			if (React.isValidElement(item)) {
				const domProps = merge(item.props, { style: item.style, className: item.className });
				return list.push(
					<React.Fragment key={_index}>{React.cloneElement(item, { ...domProps })}</React.Fragment>
				);
			}
			if (isObject(item)) {
				// 处理插件
				const _base = merge({}, mergeProps, item);

				const { labelCol, wrapperCol, offset } = _base;
				const { noLabelLayoutValue, labelLayoutValue } = getLabelLayout({ labelCol, wrapperCol, offset });

				let _plugins: YFormPluginsType = {};
				if (typeof _base.plugins === 'boolean') {
					defaultPlugin = _base.plugins;
				} else if (isObject(_base.plugins)) {
					_plugins = _base.plugins;
				}
				const {
					placeholder = defaultPlugin,
					required = defaultPlugin,
					noLabelLayout = defaultPlugin,
					labelLayout = defaultPlugin,
					disabled = defaultPlugin,
				} = _plugins;

				if ('isShow' in item && !item.isShow) {
					return undefined;
				}
				const { type, componentProps, dataSource, items, addonAfter, plugins, ...fieldRest } = item; //  list 默认不需要 FormItem 样式

				let dom;
				let _children = item.children;
				let _formItemProps = { ...fieldRest };
				if (labelLayout) {
					_formItemProps = { ..._formItemProps, ...labelLayoutValue };
				}
				let key: React.ReactText = _index;
				let _componentProps = {};
				const { label, name } = _formItemProps;
				if (disabled) {
					set(_componentProps, 'disabled', mergeDisabled);
				}
				_componentProps = { ..._componentProps, ...componentProps };
				// 添加无 label 处理
				if (noLabelLayout && !label) {
					_formItemProps = { ..._formItemProps, ...noLabelLayoutValue, labelCol: undefined };
				}
				if (item.type && itemsType) {
					const _fieldData = itemsType[item.type];
					if (_fieldData) {
						const { component, render, formatStr, formItemProps, modifyProps } = _fieldData;
						// 注入定义类型传的默认数据
						_formItemProps = { ..._formItemProps, ...formItemProps };

						//  添加必填 placeholder 处理
						if ((placeholder || required) && name) {
							const _formatStr = typeof label === 'string' && replaceMessage(formatStr || '', { label });
							if (placeholder) {
								_componentProps = { placeholder: _formatStr || '请输入', ..._componentProps };
							}
							if (required) {
								_formItemProps.rules = merge(
									[],
									[{ required: mergeRequired, message: _formatStr || '此处不能为空' }],
									_formItemProps.rules
								);
							}
						}
						// 最后修改参数
						if (modifyProps) {
							[_formItemProps, _componentProps] = modifyProps(_formItemProps, _componentProps);
						}
						const _key = name ? `${name}` : key;
						key = find(list, { key: _key }) ? key : _key;
						// 包含 items 类型把当前 item 属性全部透传过去
						if (items) {
							// list 类型需要 disabled
							_componentProps = { ...pick(_componentProps, 'disabled'), ...item, key };
						}

						if (render) {
							dom = render(_componentProps);
						} else if (component) {
							dom = React.cloneElement(component, _componentProps);
						} else if (type === 'custom') {
							dom = React.cloneElement(item.component, _componentProps);
						}
						_children = dom;
					} else {
						warning(false, `[YFom.Items] ${type} 类型未找到`);
					}
				} else {
					// 没有 type 单独有 dataSource 情况
					if (dataSource) {
						_children = (
							<Items plugins={plugins} {..._componentProps}>
								{dataSource}
							</Items>
						);
					}
				}
				// list 不需要 Form.Item
				if (type === 'list') {
					list.push(_children);
				} else {
					list.push(
						<ItemChildren key={key} addonAfter={addonAfter} {..._formItemProps}>
							{typeof _children === 'function'
								? (form: FormInstance) => {
										return (
											<Items noStyle plugins={plugins}>
												{(_children as YFormRenderChildren)(form)}
											</Items>
										);
								  }
								: _children}
						</ItemChildren>
					);
				}
			} else {
				return list.push(item);
			}
		});
	};

	if (isArray(children)) {
		each(children as YFormItemsTypeArray<InternalYFormItemProps>[]);
	} else {
		list.push(children);
	}

	const child = <YFormItemsContext.Provider value={{ ...mergeProps }}>{list}</YFormItemsContext.Provider>;
	return noStyle ? (
		child
	) : (
		<div className={classNames('yform-items', className)} style={style}>
			{child}
		</div>
	);
};

export default Items;
