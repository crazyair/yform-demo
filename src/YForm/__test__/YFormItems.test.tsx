// tslint:disable:no-console
import React from 'react';
import { render } from '@testing-library/react';
import KeyCode from 'rc-util/lib/KeyCode';
import { mount } from 'enzyme';
import { Input } from 'antd';
import { YForm } from '../../index';
import { YFormItemsProps } from '../Items';

const delay = (timeout = 0) =>
	new Promise(resolve => {
		setTimeout(resolve, timeout);
	});

const YFormItemsDemo = (props: YFormItemsProps) => {
	return (
		<YForm>
			<YForm.Items {...props}>
				{[
					{ type: 'input', name: 'demo' },
					{ type: 'money', name: 'money', label: 'money' },
					{
						dataSource: [
							{
								type: 'button',
								noStyle: true,
								plugins: { disabled: false },
								componentProps: { type: 'primary', htmlType: 'submit', children: 'submit' },
							},
						],
					},
				]}
			</YForm.Items>
		</YForm>
	);
};

describe('YFormItems', () => {
	const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

	beforeEach(() => {
		jest.useRealTimers();
	});

	afterEach(() => {
		errorSpy.mockReset();
	});

	afterAll(() => {
		errorSpy.mockRestore();
	});

	async function operate(wrapper: any, className: string) {
		wrapper
			.find(className)
			.last()
			.simulate('click');
		await delay();
		wrapper.update();
	}

	test('renders', () => {
		const wrapper = render(<YFormItemsDemo />);
		expect(wrapper).toMatchSnapshot();
	});
	test('isShow', () => {
		const wrapper = render(<YFormItemsDemo isShow={false} />);
		expect(wrapper).toMatchSnapshot();
	});
	test('no children', () => {
		const wrapper = render(
			<YForm>
				<YForm.Items />
			</YForm>
		);
		expect(wrapper).toMatchSnapshot();
	});
	test('more children', () => {
		const wrapper = mount(
			<YForm required>
				<YForm.Items plugins={{ required: true }}>
					{[
						{ type: 'input', name: 'input' },
						{ type: 'input', name: 'input_isShow', isShow: false },
						{
							type: 'list',
							name: 'list',
							items: ({ index }) => [{ type: 'input', name: [index, 'd'] }],
						},
						{
							type: 'textarea',
							name: 'textarea',
							label: '长文本',
							componentProps: { inputMax: 9 },
						},
						{
							type: 'radio',
							name: '单选框',
							componentProps: { onAddProps: () => 'a', options: [{ name: '真的', id: '1' }] },
						},
						{
							type: 'radio',
							name: '单选框',
							componentProps: { renderOption: () => 'a', options: [{ name: '真的', id: '1' }] },
						},
						{
							type: 'select',
							label: '下拉框',
							name: '下拉框',
							componentProps: {
								optionLabelProp: 'checkedValue',
								onAddProps: item => ({ checkedValue: `(${item.name})` }),
								showField: record => (
									<div>
										<div>{record.id}</div>-{record.name}
									</div>
								),
								options: [{ id: '1', name: '开' }, { id: '2', name: '关' }],
							},
						},
						{
							type: 'select',
							label: '下拉框',
							name: '下拉框',
							componentProps: {
								optionLabelProp: 'checkedValue',
								renderOption: () => 'a',
								options: [{ id: '1', name: '开' }],
							},
						},
						{ shouldUpdate: true, children: () => [{ type: 'input', name: 'f' }] },
						{ type: 'checkbox', name: 'checkbox' },
						{
							type: 'checkboxGroup',
							name: 'checkboxGroup',
							componentProps: { onAddProps: () => 'a', options: [{ id: '1', name: 'a' }] },
						},
						{
							type: 'checkboxGroup',
							name: 'checkboxGroup2',
							componentProps: { options: [{ id: '1', name: 'a' }], renderOption: () => 'b' },
						},
						{ type: 'list', items: ({ index }) => [{ type: 'input', name: [index, 'd'] }] },
						{ type: 'money', name: 'money' },
						{ label: 'text', name: 'text', type: 'text' },
						{ label: 'switch', name: 'switch', type: 'switch' },
						{ label: 'custom', type: 'custom', name: 'custom', component: <Input /> },
					]}
					{[{ type: 'noType', name: 'a' }] as any}
					<div>1</div>
					123
				</YForm.Items>
			</YForm>
		);
		expect(wrapper).toMatchSnapshot();
	});
	test('list max min', async () => {
		const wrapper = mount(
			<YForm initialValues={{ a: [1, 2, 3] }}>
				<YForm.Items>
					{[
						{
							type: 'list',
							name: 'a',
							componentProps: { maxNum: 3, minNum: 3 },
							items: ({ index }) => [{ type: 'input', name: [index] }],
						},
					]}
				</YForm.Items>
			</YForm>
		);
		expect(wrapper).toMatchSnapshot();
	});
	test('list', async () => {
		const wrapper = mount(
			<YForm initialValues={{ users: [{}, {}] }}>
				<YForm.Items>
					{[
						{
							type: 'list',
							name: 'phones',
							componentProps: {
								showIcons: {
									showBottomAdd: { text: '添加手机号' },
									showAdd: true,
									showRemove: false,
								},
								onShowIcons: () => ({ showAdd: true, showRemove: true }),
							},
							items: ({ index }) => {
								return [{ label: index === 0 && '手机号', type: 'input', name: [index, 'phone'] }];
							},
						},
					]}
				</YForm.Items>
			</YForm>
		);

		await operate(wrapper, '.ant-btn-dashed');
		expect(wrapper.find('.ant-form-item').length).toBe(4);
		await operate(wrapper, '.anticon-plus-circle');
		expect(wrapper.find('.ant-form-item').length).toBe(6);
		await operate(wrapper, '.anticon-minus-circle');
		expect(wrapper.find('.ant-form-item').length).toBe(4);
	});
	test('money', async () => {
		const wrapper = mount(
			<YForm initialValues={{ money: '1' }}>
				<YForm.Items>{[{ type: 'money', name: 'money' }]}</YForm.Items>
			</YForm>
		);
		const dom = await wrapper.find('.ant-input').last();
		dom.simulate('change', { target: { value: '2' } });
		dom.simulate('blur');
		dom.simulate('change', { target: { value: 'a' } });
		dom.simulate('blur');
	});
	test('text', async () => {
		const wrapper = mount(
			<YForm>{[{ name: 'text', type: 'text', componentProps: { editable: { onStart: () => {} } } }]}</YForm>
		);
		wrapper
			.find('.ant-typography-edit')
			.first()
			.simulate('click');
		wrapper.find('.ant-input').simulate('change', { target: { value: 'Bamboo' } });
		await wrapper.find('.ant-input').simulate('keyDown', { keyCode: KeyCode.ENTER });
		await wrapper.find('.ant-input').simulate('keyUp', { keyCode: KeyCode.ENTER });
	});
	test('textarea', async () => {
		const wrapper = mount(
			<YForm initialValues={{ textarea: '这是长文本' }}>
				{[
					{
						type: 'textarea',
						name: 'textarea',
						plugins: { required: false },
						label: '长文本',
						componentProps: { inputMax: 9 },
					},
				]}
			</YForm>
		);
		const dom = await wrapper.find('.ant-input').last();
		dom.simulate('change', { target: { value: 'a' } });
		dom.simulate('change', { target: { value: '这是长文本把' } });
	});

	test('oneLine', async () => {
		const wrapper = mount(
			<YForm>
				{[
					{
						label: '多字段',
						type: 'oneLine',
						componentProps: { oneLineStyle: ['50%', 8, '50%'] },
						plugins: { noLabelLayout: false },
						items: () => [
							{ label: '姓名', type: 'input', name: 'name' },
							<span key="center" />,
							{ label: '年龄', type: 'input', name: 'age' },
						],
					},
					{ label: '多字段', type: 'oneLine', items: () => [undefined] },
				]}
			</YForm>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
