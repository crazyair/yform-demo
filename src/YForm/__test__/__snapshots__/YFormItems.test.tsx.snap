// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YFormItems isShow 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-horizontal"
      >
        <div
          class="yform-items"
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="yform-items"
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`YFormItems list max min 1`] = `ReactWrapper {}`;

exports[`YFormItems more children 1`] = `ReactWrapper {}`;

exports[`YFormItems no children 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-horizontal"
      >
        <div
          class="yform-items"
        >
          <div
            class="yform-items"
          />
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="yform-items"
      >
        <div
          class="yform-items"
        />
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`YFormItems oneLine 1`] = `ReactWrapper {}`;

exports[`YFormItems renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-horizontal"
      >
        <div
          class="yform-items"
        >
          <div
            class="yform-items"
          >
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <input
                    class="ant-input"
                    id="demo"
                    placeholder="请输入"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class=""
                  title="money"
                >
                  money
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="input-money"
                  >
                    <span
                      class="ant-input-affix-wrapper"
                    >
                      <input
                        class="ant-input"
                        id="money"
                        placeholder="请输入money"
                        type="text"
                        value=""
                      />
                      <span
                        class="ant-input-suffix"
                      >
                        元
                      </span>
                    </span>
                    <div
                      class="zh"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="yform-items"
                  >
                    <button
                      class="ant-btn ant-btn-primary"
                      type="submit"
                    >
                      <span>
                        submit
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="yform-items"
      >
        <div
          class="yform-items"
        >
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <input
                  class="ant-input"
                  id="demo"
                  placeholder="请输入"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                title="money"
              >
                money
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="input-money"
                >
                  <span
                    class="ant-input-affix-wrapper"
                  >
                    <input
                      class="ant-input"
                      id="money"
                      placeholder="请输入money"
                      type="text"
                      value=""
                    />
                    <span
                      class="ant-input-suffix"
                    >
                      元
                    </span>
                  </span>
                  <div
                    class="zh"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="yform-items"
                >
                  <button
                    class="ant-btn ant-btn-primary"
                    type="submit"
                  >
                    <span>
                      submit
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
